test/gtest/metadata_exchange.cpp

```
@@ -0,0 +1,548 @@
+    ASSERT_EQ(status, NIXL_SUCCESS);
+
+    std::this_thread::sleep_for(sleep_time);
+
+    status = dst.agent->checkRemoteMD(src.name, valid_descs.trim());
+    ASSERT_EQ(status, NIXL_SUCCESS);
+
+    status = dst.agent->checkRemoteMD(src.name, invalid_descs.trim());
+    ASSERT_EQ(status, NIXL_ERR_NOT_FOUND);
+
+    status = src.agent->invalidateLocalMD(&send_args);
+    ASSERT_EQ(status, NIXL_SUCCESS);
+
+    std::this_thread::sleep_for(sleep_time);
+
+    // Step 3: Get and load again but with additional extra params
+
+    send_args.backends.push_back(src.backend_handle);
+    send_args.includeConnInfo = true;
+
+    status = src.agent->sendLocalPartialMD(valid_descs, &send_args);
+    ASSERT_EQ(status, NIXL_SUCCESS);
+
+    std::this_thread::sleep_for(sleep_time);
+
+    status = dst.agent->checkRemoteMD(src.name, valid_descs.trim());
+    ASSERT_EQ(status, NIXL_SUCCESS);
+
+    status = dst.agent->checkRemoteMD(src.name, invalid_descs.trim());
+    ASSERT_EQ(status, NIXL_ERR_NOT_FOUND);
+}
+
+TEST_F(MetadataExchangeTestFixture, SocketSendLocalPartialWithErrors)
+{
+    auto &src = agents[0];
+    auto &dst = agents[1];
+
+    src.createAgentBackend();
+
+    size_t count = (std::rand() % 10) + 1;
+    size_t size = (std::rand() % 1024) + 1;
+    src.initAndRegisterBuffers(count, size);
+
+    auto sleep_time = std::chrono::milliseconds(500);
+    nixl_status_t status;
+    nixl_blob_t md;
+
+    nixl_opt_args_t send_args;
+    send_args.ipAddr = dst.ip;
+    send_args.port = dst.port;
+
+    // Case 1: Use unregistered descriptors
+    MemBuffer unregistered_buffer(1024, DRAM_SEG);
+    nixl_reg_dlist_t unregistered_descs(DRAM_SEG);
+    unregistered_descs.addDesc(unregistered_buffer.getBlobDesc());
+
+    status = src.agent->sendLocalPartialMD(unregistered_descs, &send_args);
+    ASSERT_NE(status, NIXL_SUCCESS);
+
+    // Case 2: Attempt to load connection info on agent without backend
+    // TODO: This currently raises exception in commWorker thread
```

not sure what the expectation is here, if there are no backends, there are no metadata to exchange

---

Right, what I wonder is if whether we want to raise an exception on a remote agent just because the local agent sent an invalid/corrupted MD or didn't wait for the remote to create backends. Basically the local agent can easily kill the remote agent.

---

I can easily change it to just log error instead of throw. A future TODO may be to implement an ACK mechanism to notify the sender on success/failure.

---

Done